### ISSUE TITLE
Playback controls on navigable renderings

### ISSUE THREAD

**Issue by Cybernide** [Shared-Reality-Lab/IMAGE-browser#141]
Combining object detection and semantic segmentation turned out to be a great idea. Loving that we can navigate between all the segments and objects seamlessly, but now we've got the problem where we can't pause or stop playback easily. From a UX perspective, we should have media controls available. Is this possible somehow? @JRegimbal letting you know so that, at your earliest convenience, we can have a talk about how to do this.

**Comment by JRegimbal** [comment]
Absolutely possible with some rewriting. 

https://github.com/Shared-Reality-Lab/IMAGE-browser/blob/11d44396adf2e16d41c391d85a2f7f61d5fbde39/src/info/info.ts#L166-L169

We create a new [AudioBufferSourceNode](https://developer.mozilla.org/en-US/docs/Web/API/AudioBufferSourceNode) whenever the play button is pressed. The code could be refactored so that we have one possibly undefined node that can then be initialized and started using the play button and stopped using a new stop button. Pausing may require us to add some logic tracking the time it was stopped at (I don't know if we get that for free - I'm not an expert with WebAudio) but stopping is just using the [AudioScheduledSourceNode.stop](https://developer.mozilla.org/en-US/docs/Web/API/AudioScheduledSourceNode/stop) method.

**Comment by JRegimbal** [comment]
It looks like if we have per-rendering audio context objects it gets easier [based on this explanation for pausing](https://github.com/WebAudio/web-audio-api-v2/issues/105#issuecomment-723691179). Then we can just suspend/resume the entire [context](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext) without impacting the source node.

**Comment by Cybernide** [comment]
So @Clarisa1999 or maybe @jaydeepsingh25 Can we add playback controls to the individual parts of the renderings like so?
![image](https://user-images.githubusercontent.com/7490996/152004939-b26e1dcf-4778-4f81-95af-1851f425d9c0.png)

**Comment by Cybernide** [comment]
@Sabrina-Knappe mentioned the possibility of making this even more screenreader accessible, having a chat with her and then updating this issue.

**Comment by Cybernide** [comment]
@Sabrina-Knappe You seem to have some ideas as to how to go about this to ensure decent functionality between all screen readers. Please let me know when you have a flowchart/diagram/wireframe to review.

**Comment by Sabrina-Knappe** [comment]
![IMG_1151](https://user-images.githubusercontent.com/56092221/152603961-4d436278-d979-4981-8662-7ee6b8c89b9f.JPG)
Here is a very rough state diagram for @JRegimbal

**Comment by JRegimbal** [comment]
This seems like it should work...I suppose we aren't pausing after the end of a segment? Still a bit nervous about the visibility of the system but honestly worst case this is just something we can fix by inserting HTML comments.

Do we want this piggybacking on the plyr display? And if so, is it worth worrying about changing the behavior of seeking?

**Comment by jeffbl** [comment]
From triage:
 - @JRegimbal implements hack to at least allow pause/stop
 - Then goes to @Cybernide  for final design review of full implementation with extension team to implement by CSUN

**Comment by Sabrina-Knappe** [comment]
@JRegimbal I'm not sure whether pausing at the end of each segment makes more sense or not... It matches the current behaviour of the dropdown more closely...

**Comment by JRegimbal** [comment]
Ok. And are we aiming to ultimately integrate into plyr?

**Comment by jeffbl** [comment]
@Sabrina-Knappe My opinion is that it should go straight through unless the user selects otherwise, since that gives the full "experience". If they want it broken up, they can still use the dropdown, correct?

**Comment by JRegimbal** [comment]
@jeffbl I thought that the plan here was to *remove* the dropdown since it caused problems with voiceover.

**Comment by Sabrina-Knappe** [comment]
Yes, the idea was to remove the dropdown. Would wireframes be useful for this?

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by jeffbl** [comment]
Apologies - I hadn't gone through the whole conversation, and I missed that. VoiceOver compatibility is important. I'll just note that iirc one participant seemed to like the dropdown since you can select the exact thing you want to hear, but if this is very fast skipping forward/back, probably no more clicking/listening than navigating the dropdown. RE: playing straight through, my suspicion is that most often, the desire will be to hear the entire rendering without having to push more buttons.

**Comment by Cybernide** [comment]
Heads up, I just had a discussion with the haptics crew, and they're looking for guidance on the UI for haptics - I told them that I'm waiting to see what we decide on in this issue, as I want to ensure that there's some consistency between rendering types.

**Comment by JRegimbal** [comment]
Looking for guidance on this as there were problems when used with VoiceOver (that, IIRC, were also experienced by players like YouTube). I only have access to Orca on my development machine, where it does work when set to brief rather than verbose mode, so I need feedback on what changes are necessary and ideally how to implement for compatibility with the big screen readers.

**Comment by Sabrina-Knappe** [comment]
@JRegimbal I am looking at alternatives that don't involve using a player. Would be similar to what happens with current dropdown where the button triggers the playing of a section of the clip, but without the dropdown structure. 
@jeffbl It's important to note that the navigable UI is functional in the current main, it's just a bit annoying to use. You should be okay passing this to CSUN if necessary, imo.

**Comment by jeffbl** [comment]
@JRegimbal My understanding is that we still don't have a solid proposal here for full playback controls, so I am postponing this to post-CSUN. If there is a lighter-weight change going in that still has work to do (e.g., just a "stop" control), please open a separate issue for that.

**Comment by JRegimbal** [comment]
We don't have a solid proposal yet. @Sabrina-Knappe has been working on an idea that seems promising, but I don't think has been fully vetted yet (correct me if I'm wrong). We have modified playback controls in as part of the rendering playback overlapping issue.

**Comment by Sabrina-Knappe** [comment]
It needs to be vetted on NVDA/JAWS, then it is ready to be implemented.

From: Juliette Regimbal ***@***.***>
Date: Friday, February 25, 2022 at 3:39 PM
To: Shared-Reality-Lab/IMAGE-browser ***@***.***>
Cc: Sabrina Knappe ***@***.***>, Mention ***@***.***>
Subject: Re: [Shared-Reality-Lab/IMAGE-browser] Playback controls on navigable renderings (Issue #141)

We don't have a solid proposal yet. @Sabrina-Knappe<https://github.com/Sabrina-Knappe> has been working on an idea that seems promising, but I don't think has been fully vetted yet (correct me if I'm wrong). We have modified playback controls in as part of the rendering playback overlapping issue.

**Comment by JRegimbal** [comment]
@Sabrina-Knappe any update on the testing for this?

**Comment by Sabrina-Knappe** [comment]
Works on NVDA. Cyan’s JAWS expired so I’m going to try to get JAWS and test with it.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by JRegimbal** [comment]
Reassigning to @Sabrina-Knappe until it's vetted and there's something for me to actually test/implement!

**Comment by Cybernide** [comment]
@jeffbl Recommend postponing - UX team will discuss what needs to be done

**Comment by Sabrina-Knappe** [comment]
Gated by #168 
We need to test the new functionality to find out how to proceed.

**Comment by jeffbl** [comment]
Moving to next sprint because #168 is in testing, and this is gated on that testing...

**Comment by Sabrina-Knappe** [comment]
This is temporarily solved by the new "button" layout.
Further investigation in #272

**Comment by igfx83** [comment]
//I achieve this functionality by extending the AudioBufferSourceNode with a "processedSamples" property:
 public async setupAudioNodes(usedSource?: any): Promise<AudioBufferSourceNode> {
      const newSource = this.audioContext.createBufferSource() as any;
  
//When I make instantiate a new ABSN I omit the param and init at zero. When the node is recycled I include it as the param and get the //processed Samples value:    
      Object.assign(newSource, {
        processedSamples: usedSource ? usedSource.processedSamples : 0,
      });
      
 //I create an AudioWorklet if one hasn't been created already:
      if(!this.audioProcessor) this.audioProcessor = new AudioWorkletNode(this.audioContext, 'sampler');

//And listen to it's message port. The processor .js just at some point counts every sample and emits the count after each frame:
      this.audioProcessor.port.onmessage = (event: MessageEvent) => {
        newSource.processedSamples += event.data;
      };

//I run it through the rest of the nodes and to audio out:
      newSource.connect(this.audioProcessor, 0, 0);
      this.audioProcessor.connect(this.audioContext.destination, 0, 0);

return newSource;
}

//Elsewhere register a listener for the sourceNodes "onEnded" event before sourceNode.start() is called. The handler contains the logic that  recycles the node while maintaining the sample count:
      sourceNode.onEnded( async () => {
             const newSource: AudioBufferSourceNode = await this.AudioService.setupAudioNodes(this.sourceNode);
             const buffer: AudioBuffer = this.sourceNode.buffer;
             newSource.buffer = buffer;
             sourceNode = newSource;
});

//When play is called, the offset timing is adjusted using  processedSamples / sampleRate :
       public playAudio(): void {
              const offset: number = this.sourceNode.processedSamples / this.sourceNode.context.sampleRate;
              sourceNode.start(0, offset);
}

//This seperates the SourceNodes from the Audio Context, allowing other nodes on the Context to continue to exist while maintaining seamless repeatability with sufficient accuracy.
